pytorch/torchx
# slurm_scheduler: support retries in slurm and better handle failures

Closes #479
Closes #394 

This makes our slurm scheduler integration support max_retries the same way that our other schedulers do. This only handles whole job level retries and doesn't support per replica retries.

See Also:
* requeuing job on failure https://stackoverflow.com/a/42773008/4722305
* https://slurm.schedmd.com/sbatch.html#SECTION_OUTPUT-ENVIRONMENT-VARIABLES

Test plan:


CI + integ tests

Will test by launching a failing job on our slurm cluster and make sure it dies after N retries

```
(tristanr) ubuntu@ip-10-0-0-161:~/tristanr$ cat slurm-2593.out 
+ export PYTHONUNBUFFERED=1
+ PYTHONUNBUFFERED=1
+ export SLURM_UNBUFFEREDIO=1
+ SLURM_UNBUFFEREDIO=1
+ export TORCHX_MAX_RETRIES=5
+ TORCHX_MAX_RETRIES=5
+ set +e
+ srun --output=slurm-2593-sh-0.out --error=slurm-2593-sh-0.err --wait=60 --kill-on-bad-exit=1 --export=ALL,LOGLEVEL=WARNING sh -c ./env-fail.sh : --output=slurm-2593-sh-1.out --error=slurm-2593-sh-1.err --wait=60 --kill-on-bad-exit=1 --export=ALL,LOGLEVEL=WARNING sh -c ./env-fail.sh
srun: error: compute-dy-t3small-1: task 0: Exited with exit code 1
srun: Terminating job step 2594.1
srun: Terminating job step 2593.1
srun: error: compute-dy-t3small-2: task 0: Terminated
srun: Force Terminated job step 2594.1
+ exitcode=143
+ set -e
+ echo 'job exited with code 143'
job exited with code 143
+ '[' 143 -ne 0 ']'
+ '[' 5 -gt 1 ']'
+ scontrol requeue 2593
+ exit 143
slurmstepd: error: *** JOB 2593 ON compute-dy-t3small-1 CANCELLED AT 2022-05-20T21:52:32 DUE TO JOB REQUEUE ***
(tristanr) ubuntu@ip-10-0-0-161:~/tristanr$ squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) 
            2593+0   compute     sh-0   ubuntu PD       0:00      1 (BeginTime) 
            2593+1   compute     sh-1   ubuntu PD       0:00      1 (BeginTime) 
(tristanr) ubuntu@ip-10-0-0-161:~/tristanr$ squeue --Format JobId,Nodelist,Name,RestartCnt
JOBID               NODELIST            NAME                RESTART_COUNT       
2593                                    sh-0                2                   
2594                                    sh-1                2                   
```
